libs/SmartSync/SmartSync/Classes/Manager/SFLayoutSyncManager.m

```diff
@@ -0,0 +1,162 @@
+        [syncMgrList removeAllObjects];
+    }
+}
+
++ (void)reset:(SFUserAccount *)user {
+    @synchronized([SFLayoutSyncManager class]) {
+        if (user) {
+            NSString *matchingKey = SFKeyForUserAndScope(user, SFUserAccountScopeCommunity);
+            NSArray<NSString *> *keys = syncMgrList.allKeys;
+            for (NSString *key in keys) {
+                if ([key hasPrefix:matchingKey]) {
+                    [syncMgrList removeObjectForKey:key];
+                }
+            }
+        }
+    }
+}
+
++ (void)initialize {
+    if (self == [SFLayoutSyncManager class]) {
+        syncMgrList = [NSMutableDictionary new];
+        indexSpecs = [NSArray arrayWithObjects:[[SFSoupIndex alloc] initWithPath:@"sobjectType" indexType:kSoupIndexTypeJSON1 columnName:@"sobjectType"],
+                       [[SFSoupIndex alloc] initWithPath:@"layoutType" indexType:kSoupIndexTypeJSON1 columnName:@"layoutType"], nil];
+    }
+}
+
+- (void)fetchLayoutForObject:(NSString *)objectType layoutType:(NSString *)layoutType mode:(SFSDKFetchMode)mode completionBlock:(SFLayoutSyncCompletionBlock)completionBlock {
+    switch (mode) {
+        case SFSDKFetchModeCacheOnly:
+            [self fetchFromCache:objectType layoutType:layoutType completionBlock:completionBlock fallbackOnServer:NO];
+            break;
+        case SFSDKFetchModeCacheFirst:
+            [self fetchFromCache:objectType layoutType:layoutType completionBlock:completionBlock fallbackOnServer:YES];
+            break;
+        case SFSDKFetchModeServerFirst:
+            [self fetchFromServer:objectType layoutType:layoutType completionBlock:completionBlock];
+            break;
+    }
+}
+
+- (instancetype)init:(SFSmartSyncSyncManager *)syncManager {
+    self = [super init];
+    if (self) {
+        self.syncManager = syncManager;
+        self.smartStore = syncManager.store;
+        [self initializeSoup];
+    }
+    return self;
+}
+
+- (void)fetchFromServer:(NSString *)objectType layoutType:(NSString *)layoutType completionBlock:(SFLayoutSyncCompletionBlock)completionBlock {
+    SFLayoutSyncDownTarget *target = [SFLayoutSyncDownTarget newSyncTarget:objectType layoutType:layoutType];
+    __weak typeof (self) weakSelf = self;
+    [self.syncManager syncDownWithTarget:target soupName:kSoupName updateBlock:^(SFSyncState *sync) {
+        __strong typeof (weakSelf) strongSelf = weakSelf;
+        if (sync.status == SFSyncStateStatusDone) {
+            [strongSelf fetchFromCache:objectType layoutType:layoutType completionBlock:completionBlock fallbackOnServer:NO];
+        }
+    }];
+}
+
+- (void)fetchFromCache:(NSString *)objectType layoutType:(NSString *)layoutType completionBlock:(SFLayoutSyncCompletionBlock)completionBlock fallbackOnServer:(BOOL)fallbackOnServer {
+    SFQuerySpec *querySpec = [SFQuerySpec newSmartQuerySpec:[NSString stringWithFormat:kQuery, kSoupName, kSoupName, kSoupName, objectType, kSoupName, layoutType] withPageSize:1];
```

Comment:
With the page size 1, if there are multiple rows for the same layout, we won't notice.
Before today's fix, it was probably happening on Android.
In the tests, could we have one that does multiple fetchFromServer in a row, followed by a select count(*) on the table?